Add soft limits as appendix C (#256)
oasis-tcs/csaf: csaf_2.0/prose/csaf-v2-editor-draft.md

    cisco-sa-20190513-secureboot
```

> This value is also used to define the filename for the CSAF document. The following rules MUST be applied to determine the filename for the CSAF document:

Reviewer: We specify the filename of our CSAF document in a quite hidden section under `Document Property - Tracking - ID`. I would not have expected that as a reader. Can we find some other, more adequate place for this specification so that it can also be found in the TOC?

Author: We should track that as a separate issue. Therefore: See #277.

Informatiove tests provide insights in common mistakes and bad practices. They MAY fail at a valid CSAF document. It is up to the issuing party to decide whether this was an intended behavior and can be ignore or should be treated. These tests may include information about recommended usage. A program MUST handle a test failure as a information.

# 5 Distributing CSAF documents

This section lists requirements and roles defined for distributing CSAF documents. The first subsection provides all requirements - the second one the roles. It is mandatory to fulfill the basic role "CSAF publisher".

## 5.1 Requirements

The requirements in this subsection are consecutively numbered to be able to refer to them directly. The order does not give any hint about the importance.

### 5.1.1 Requirement 1: Valid CSAF document

The document is a valid CSAF document (cf. Conformance clause 1).

### 5.1.2 Requirement 2: Filename

The CSAF document has a filename according to the rules in section 3 (cf. section 3.2.1.12.4).

### 5.1.3 Requirement 3: TLS

> The CSAF document is retrievable from a website which uses TLS for encryption and server authenticity. The CSAF document MUST not be downloadable from a location which does not encrypt the transport.

Reviewer: I do not understand why we put emphasis on encryption here. In many cases the document will be public, so we should focus on server authentication and data integrity. 
Do we want to also force TLS? What about PGP to ensure those requirements? I am thinking about the typical package mirrors of Linux distributions that are still available via HTTP and FTP.

| Revision | Date | Editor | Changes Made |
| :--- | :--- | :--- | :--- |
| csaf-v2.0-wd20210521 | 2021-05-21 | Stefan Hagen and Thomas Schmidt| Editor revision for TC review |

# Appendix C. Guidance on the Size of CSAF Documents

The TC carefully considered all known aspects to provide size limits for CSAF documents for this version of the specification. It was decided that hard limits should not be enforced. However, since there is the need for guidance to ensure interoperability in the eosystem, the TC provides a set of soft limits. A CSAF document which exceeds those, can still be valid but it might not be processable for some parties.

All _CSAF consumers_ should be able to process CSAF documents which comply with the limits below. All _CSAF producers_ should not produce CSAF documents which exceed those limits.

> If you come across a case where these limits are exceeded, please provide feedback to the TC.

## File size

A CSAF document in the specified JSON format encoded in UTF-8 should be not lager than 15 MB.

> >At least one database technology in wide use for storing CSAF documents rejects insert attempts when the transformed BSON size exceeds 16 megabytes. The BSON format optimizes for accessibility and not size. So, small integres and small strings may incur more overhead in the BSOn format than in JSON. In addition, the BSON format adds lebgth information for the entries inside the document which adds to the size when storing CSAF document content in a BSON format.

Reviewer 2: ```suggestion
> At least one database technology in wide use for storing CSAF documents rejects insert attempts when the transformed BSON size exceeds 16 megabytes. The BSON format optimizes for accessibility and not size. So, small integers and small strings may incur more overhead in the BSON format than in JSON. In addition, the BSON format adds length information for the entries inside the document, which adds to the size when storing CSAF document content in a BSON format.
```

  * `/product_tree/branches[]/product/product_identification_helper/cpe`
  * `/product_tree/branches[]/product/product_identification_helper/purl`
  * `/product_tree/branches[](/branches[])*/product/product_identification_helper/cpe`
  * `/product_tree/branches[](/branches[])*/product/product_identification_helper/purl`
  * `/product_tree/full_product_names[]/product_identification_helper/cpe`
  * `/product_tree/full_product_names[]/product_identification_helper/purl`
  * `/product_tree/product_groups[]/summary`
  * `/product_tree/relationships[]/full_product_name/product_identification_helper/cpe`
  * `/product_tree/relationships[]/full_product_name/product_identification_helper/purl`
  * `/vulnerabilities[]/acknowledgments[]/summary`
  * `/vulnerabilities[]/involvements[]/summary`
  * `/vulnerabilities[]/references[]/summary`
  * `/vulnerabilities[]/remediations[]/entitlements[]`

* 30 000 for
  * `/document/notes[]/text`
  * `/vulnerabilities[]/notes[]/text`

* 250 000 for
  * `/vulnerabilities[]/remediations[]/details`
  * `/vulnerabilities[]/remediations[]/restart_required/details`
  * `/vulnerabilities[]/threats[]/details`

## URI length

> A string with format `uri` should not have a length greater than 20000. This applies to:

Reviewer 2: Most browsers have a limit of 2,000 characters on URLs, so 20,000 seems very generous.

Reviewer 3: Well, that is a URI :wink: not necessarily a URL ... but, the limits we will publish after discussing specific first (possibly other) limits will be the real thing and gain of this nice proposal from @tschmidtb51 ... :+1:

Reviewer 2: True, but if you look at the list of attributes below, more than a third are URLs even though the type of the field is technically `uri`.

* `/document/acknowledgments[]/urls[]`
* `/document/aggregate_severity/namespace`
* `/document/distribution/tlp/url`
* `/document/references[]/url`
* `/product_tree/branches[]/product/product_identification_helper/x_generic_uris[]/namespace`
* `/product_tree/branches[]/product/product_identification_helper/x_generic_uris[]/uri`
* `/product_tree/branches[](/branches[])*/product/product_identification_helper/x_generic_uris[]/namespace`
* `/product_tree/branches[](/branches[])*/product/product_identification_helper/x_generic_uris[]/uri`
* `/product_tree/full_product_names[]/product_identification_helper/x_generic_uris[]/namespace`
* `/product_tree/full_product_names[]/product_identification_helper/x_generic_uris[]/uri`
* `/product_tree/relationships[]/full_product_name/product_identification_helper/x_generic_uris[]/namespace`
* `/product_tree/relationships[]/full_product_name/product_identification_helper/x_generic_uris[]/uri`
* `/vulnerabilities[]/acknowledgments[]/urls[]`
* `/vulnerabilities[]/references[]/url`
* `/vulnerabilities[]/remediations[]/url`

## Enum

A string which is an enum has a fixed maximum length given by its longest value.

> > Later versions of CSAF migth add, modify or delete possible value. Therefore, this sizes should not be implemented as fixed limits.

Reviewer 2: Not sure I understand this note and this section in general. The maximum size of a string that can use only one of the values from an enum is not variable. So as long as the CSAF spec does not exceeds the string soft limit for a particular enum, then there is nothing to adhere to as far as consumers/producers go.

Reviewer 3: Yes, I think this is maybe "belt and braces" style - I understood this information as: Look, we have now a specific set of enumerations i.e. strings with a certain fixed length as of CSAF v2.0 - do not implement a filter rejecting longer entries when you are not willing to adjust that limit later (future revisions).

I can imagine that we will have a different wording "here" when we publish.

Reviewer 2: Ah, ok. That makes sense. Perhaps rewording it to how you described it would help make that a bit clearer :-)

Author: I hope the new wording makes it clearer...

Author: @mprpic: Is it now clearer?

* `/vulnerabilities[]/scores[]/cvss_v3/availabilityImpact`
* `/vulnerabilities[]/scores[]/cvss_v3/baseSeverity`
* `/vulnerabilities[]/scores[]/cvss_v3/exploitCodeMaturity`
* `/vulnerabilities[]/scores[]/cvss_v3/remediationLevel`
* `/vulnerabilities[]/scores[]/cvss_v3/reportConfidence`
* `/vulnerabilities[]/scores[]/cvss_v3/temporalSeverity`
* `/vulnerabilities[]/scores[]/cvss_v3/confidentialityRequirement`
* `/vulnerabilities[]/scores[]/cvss_v3/integrityRequirement`
* `/vulnerabilities[]/scores[]/cvss_v3/availabilityRequirement`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedAttackVector`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedAttackComplexity`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedPrivilegesRequired`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedUserInteraction`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedScope`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedConfidentialityImpact`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedIntegrityImpact`
* `/vulnerabilities[]/scores[]/cvss_v3/modifiedAvailabilityImpact`
* `/vulnerabilities[]/scores[]/cvss_v3/environmentalSeverity`
* `/vulnerabilities[]/threats[]/category`

## Date

> The maxima length of strings representing a temporal value is given by the format specifier. This applies to:

Reviewer 2: ```suggestion
The maximum length of strings representing a temporal value is given by the format specifier. This applies to:
```

| Revision | Date | Editor | Changes Made |
| :--- | :--- | :--- | :--- |
| csaf-v2.0-wd20210521 | 2021-05-21 | Stefan Hagen and Thomas Schmidt| Editor revision for TC review |

# Appendix C. Guidance on the Size of CSAF Documents

> The TC carefully considered all known aspects to provide size limits for CSAF documents for this version of the specification. It was decided that hard limits should not be enforced. However, since there is the need for guidance to ensure interoperability in the eosystem, the TC provides a set of soft limits. A CSAF document which exceeds those, can still be valid but it might not be processable for some parties.

Reviewer: ```suggestion
The TC carefully considered all known aspects to provide size limits for CSAF documents for this version of the specification. It was decided that hard limits should not be enforced. However, since there is the need for guidance to ensure interoperability in the ecosystem, the TC provides a set of soft limits. A CSAF document which exceeds those, can still be valid but it might not be processable for some parties.
```

All _CSAF consumers_ should be able to process CSAF documents which comply with the limits below. All _CSAF producers_ should not produce CSAF documents which exceed those limits.

> If you come across a case where these limits are exceeded, please provide feedback to the TC.

## File size

> A CSAF document in the specified JSON format encoded in UTF-8 should be not lager than 15 MB.

Reviewer: ```suggestion
A CSAF document in the specified JSON format encoded in UTF-8 should not be larger than 15 MB.
```

Author: We changed that to "be smaller".
